From 90db90c72021bcc860e944b4956565b04aa56864 Mon Sep 17 00:00:00 2001
From: Liu Shixin <liushixin2@huawei.com>
Date: Sun, 20 Mar 2022 10:48:21 +0800
Subject: [PATCH] mm/dynamic_hugetlb: set/clear HPageFreed
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 90db90c72021bcc860e944b4956565b04aa56864
Modified-by-SEL: No


hulk inclusion
category: bugfix
bugzilla: 46904 https://gitee.com/openeuler/kernel/issues/I4Y0XO

--------------------------------

Patch ("mm: hugetlb: fix a race between freeing and dissolving the page") add
PageHugeFreed to check whether a page is freed in hugetlb.
Patch ("hugetlb: convert PageHugeFreed to HPageFreed flag") convert it to
HPageFreed. We need to clear it when alloc hugepage from hugetlb to and set it
when free hugepage back to hugetlb.

Signed-off-by: Liu Shixin <liushixin2@huawei.com>
Reviewed-by: Kefeng Wang <wangkefeng.wang@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 mm/dynamic_hugetlb.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/mm/dynamic_hugetlb.c b/mm/dynamic_hugetlb.c
index bec6ff560e37..7dc1d7643a35 100644
--- a/mm/dynamic_hugetlb.c
+++ b/mm/dynamic_hugetlb.c
@@ -739,6 +739,7 @@ static int alloc_hugepage_from_hugetlb(struct dhugetlb_pool *hpool,
 		if (ret)
 			continue;
 
+		ClearHPageFreed(page);
 		list_move_tail(&page->lru, &hpages_pool->hugepage_freelists);
 		h->free_huge_pages--;
 		h->free_huge_pages_node[nid]--;
@@ -780,6 +781,7 @@ static int free_hugepage_to_hugetlb(struct dhugetlb_pool *hpool)
 		set_compound_page_dtor(page, HUGETLB_PAGE_DTOR);
 
 		nid = page_to_nid(page);
+		SetHPageFreed(page);
 		list_move(&page->lru, &h->hugepage_freelists[nid]);
 		hpool->total_huge_pages--;
 		hpages_pool->free_normal_pages--;
-- 
2.26.2

